libmamba/include/mamba/util/iterator.hpp

```diff
@@ -340,13 +370,17 @@ namespace mamba::util
 
         void next_valid_iterator()
         {
-            while (m_iter != m_end && !m_pred(*m_iter))
+            while (m_iter != m_end && !(m_pred.value()(*m_iter)))
             {
                 ++m_iter;
             }
         }
 
-        Predicate m_pred;
+        // Trick to enable move and copy assignment: since lambdas are
+        // not assignable, we encpsulate them in an std::optional and
+        // rely on it to implement assignment operators. The optional
+        // should be replaced with a dedicated wrapper.
```

Comment:
Also note that most std::fucntion implementations actually use small-buffer optimization so while std::function is heavyweight and expensive for big callable objects, it should be fine for lambdas which dont capture big objects by value, so most lambdas - as  in dynamic allocations will be rare. But optional is indeed not allocating at all as it's based on the real type of the predicate.

Reply:
Also it avoids pulling the <functional> header. And I really plan to replace `std::optional` with a dedicated wrapper.